data/field_type_enum.go

```diff
@@ -0,0 +1,156 @@
+package data
+
+// this supports the enum type
+// it is diffent than the rest since it is backed by
+// a uint16, but has special semantics and interacts with the metadata
+// Unlike the other fields it can not be easily generated
+
+// NOTE: no implementation for newEnumVectorWithValues -- since
+
+type enumVector []uint16
+
+func newEnumVector(n int) *enumVector {
+	v := enumVector(make([]uint16, n))
+	return &v
+}
+
+func (v *enumVector) Set(idx int, i interface{}) {
+	(*v)[idx] = i.(uint16)
+}
+
+func (v *enumVector) SetConcrete(idx int, i interface{}) {
+	v.Set(idx, i)
+}
+
+func (v *enumVector) Append(i interface{}) {
+	*v = append(*v, i.(uint16))
+}
+
+func (v *enumVector) At(i int) interface{} {
+	return (*v)[i]
+}
+
+func (v *enumVector) PointerAt(i int) interface{} {
+	return &(*v)[i]
+}
+
+func (v *enumVector) Len() int {
+	return len(*v)
+}
+
+func (v *enumVector) CopyAt(i int) interface{} {
+	return (*v)[i]
+}
+
+func (v *enumVector) ConcreteAt(i int) (interface{}, bool) {
+	return v.At(i), true
+}
+
+func (v *enumVector) Type() FieldType {
+	return vectorFieldType(v)
+}
+
+func (v *enumVector) Extend(i int) {
+	*v = append(*v, make([]uint16, i)...)
+}
+
+func (v *enumVector) Insert(i int, val interface{}) {
```

Comment:
should this also check that the `val` does not exceed the length of defined enum variants?

Reply:
whatever does that needs to have access to the fieldConfig also -- i think the best way to do this is to have a helper/wrapper that manages both, but the underlying implementation (this) needs to operate without any schema info beyod type 😬